File: Week_02/94/binary_tree_inorder_traversal3.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        while root is not None:
            if root.left is None:
                res.append(root.val)
                root = root.right
            else:
                pre = root.left
                while pre.right is not None:
                    pre = pre.right

                pre.right = root
                root = root.left
                pre.right.left = None
        return res

    def test(self):
        n1 = TreeNode(1)
        n2 = TreeNode(2)
        n3 = TreeNode(3)

        n1.right = n2
        n2.left = n3

        ans = self.inorderTraversal(n1)
        assert ans == [1, 3, 2], f"ans: {ans}"
        print("well done")


if __name__ == "__main__":
    Solution().test()

File: Week_02/94/binary_tree_inorder_traversal3.go
package main

func inorderTraversal3(root *TreeNode) []int {
	res := make([]int, 0)
	for root != nil {
		if root.Left == nil {
			res = append(res, root.Val)
			root = root.Right
		} else {
			pre := root.Left
			for pre.Right != nil {
				pre = pre.Right
			}
			pre.Right = root
			root = root.Left
			pre.Right.Left = nil
		}
	}
	return res
}
